pkg/credentialprovider/plugin/plugin_test.go

```diff
@@ -0,0 +1,182 @@
+					Duration: f.cacheDuration,
+				},
+				AuthConfig: criauth.AuthConfig{
+					Username: f.username,
+					Password: f.password,
+				},
+			},
+		},
+	}, nil
+}
+
+func Test_Provide(t *testing.T) {
+	testcases := []struct {
+		name           string
+		pluginProvider *pluginProvider
+		image          string
+		dockerconfig   credentialprovider.DockerConfig
+	}{
+		{
+			name: "exact image match",
+			pluginProvider: &pluginProvider{
+				matchURLs: []string{"test.registry.io"},
+				cache:     cache.NewExpirationStore(cacheKeyFunc, &cacheExpirationPolicy{}),
+				plugin: &fakeExecPlugin{
+					matchKey: "test.registry.io",
+					username: "user",
+					password: "password",
+				},
+			},
+			image: "test.registry.io",
+			dockerconfig: credentialprovider.DockerConfig{
+				"test.registry.io": credentialprovider.DockerConfigEntry{
+					Username: "user",
+					Password: "password",
+					Email:    "plugin@criauth.k8s.io",
+				},
+			},
+		},
+		{
+			name: "wild card image match",
+			pluginProvider: &pluginProvider{
+				matchURLs: []string{"*.registry.io"},
+				cache:     cache.NewExpirationStore(cacheKeyFunc, &cacheExpirationPolicy{}),
+				plugin: &fakeExecPlugin{
+					matchKey: "*.registry.io",
+					username: "user",
+					password: "password",
+				},
+			},
+			image: "test.registry.io",
+			dockerconfig: credentialprovider.DockerConfig{
+				"*.registry.io": credentialprovider.DockerConfigEntry{
+					Username: "user",
+					Password: "password",
+					Email:    "plugin@criauth.k8s.io",
+				},
+			},
+		},
+		{
+			name: "multiple wild card image match",
+			pluginProvider: &pluginProvider{
+				matchURLs: []string{"*.test.*.registry.io"},
+				cache:     cache.NewExpirationStore(cacheKeyFunc, &cacheExpirationPolicy{}),
+				plugin: &fakeExecPlugin{
+					matchKey: "*.test.*.registry.io",
+					username: "user",
+					password: "password",
+				},
+			},
+			image: "foo.test.bar.registry.io",
+			dockerconfig: credentialprovider.DockerConfig{
+				"*.test.*.registry.io": credentialprovider.DockerConfigEntry{
+					Username: "user",
+					Password: "password",
+					Email:    "plugin@criauth.k8s.io",
+				},
+			},
+		},
+	}
+
+	for _, testcase := range testcases {
+		t.Run(testcase.name, func(t *testing.T) {
+			dockerconfig := testcase.pluginProvider.Provide(testcase.image)
+			if !reflect.DeepEqual(dockerconfig, testcase.dockerconfig) {
+				t.Logf("actual docker config: %v", dockerconfig)
+				t.Logf("expected docker config: %v", testcase.dockerconfig)
+				t.Error("unexpected docker config")
+			}
+		})
+	}
+}
+
+func Test_encodeRequest(t *testing.T) {
+	request := &criauth.CRIAuthPluginRequest{
+		Image: "test.registry.io/foobar",
+	}
+
+	expectedData := []byte(`{"kind":"CRIAuthPluginRequest","apiVersion":"criauth.k8s.io/v1alpha1","image":"test.registry.io/foobar"}
+`)
+	data, err := encodeRequest(request)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if !reflect.DeepEqual(data, expectedData) {
```

Comment:
Because `data` and `expectedData` is type `[]byte`. I could convert to string and compare directly, but I don't see a compelling reason to do that at this point.

Reply:
oh, I missed this. Thank you